Add dumpdata --yaml-collection-style option

matthijskooijman opened the pull request:
By default, pyyaml uses the "block" style (separate line per element)
for collections that contain nested collections, and the "flow" style
(single line for all elements) for other collections. However, when
creating fixtures or otherwise hand-editing or versioning the result of
dumpdata, the block style is preferred.

This commit adds a new option for the `dumpdata` command which allows
forcing the use of either the 'block' or 'flow' style. The default is
'auto' and thus unchanged.

This fixes [#11927](https://code.djangoproject.com/ticket/11927).

matthijskooijman commented:
```
 -   fields: {headline: Poker on TV is great!, pub_date: ! '2006-06-16 11:00:00'}
+   fields: {headline: Poker on TV is great!, pub_date: !!timestamp '2006-06-16 11:00:00'}
```

There seems to be one test failure on Windows, which apparently serializes timestamps differently in the flow style (using `!` rather than `!!timestamp`). Not sure how to fix this, it seems no other tests run into this so far. Using a model without a timestamp would be easy, but I'm not sure if any is available in the existing fixtures.

timgraham commented:
Adding format-specific options to dumpdata doesn't look like a proper separation of concerns. Instead I imagine something similar to `manage.py test` where test runners have an `add_arguments()` method that the command invokes.

matthijskooijman commented:
> Adding format-specific options to dumpdata doesn't look like a proper separation of concerns

Yeah, that part smells a bit.

>  Instead I imagine something similar to manage.py test where test runners have an add_arguments() method that the command invokes.

That makes sense and is probable feasible. To make that work, it should pre-scan the commandline for the `--format` command, just like [test prescans for `--testrunner`](https://github.com/django/django/blob/master/django/core/management/commands/test.py#L15-L26). One caveat is that `manage.py dumpdata --help` will not show the extra options, while `manage.py dumpdata --format yaml --help` will, but that is probably something that can be fixed with documentation (perhaps add a note the `--help` output).

I'll have a go at implementing this, but it might not be until the end of august due to the holidays.

matthijskooijman commented:
I had another look at this, but found a bug in the commandline-prescanning used by `manage.py test` that I wanted to mimic. I fixed that in #10307. If that approach looks ok, I'll generalize `prescan_testrunner_option()` and move it into `django.core.management.utils` to be reused by `manage.py dumpdata` to prescan the `--format` option.

matthijskooijman commented:
I went ahead and implemented the format-specific arguments for dumpdata as well. I now have something that works and is reasonably elegant, except that it does not work for the test suite (more on that below). I just pushed new commits that:
 - Include the `manage.py test` bugfix related to testrunner-specific arguments mentioned above. It also includes a new commit that improves the `--help` output related to the testrunner-specific arguments.
 - Adds prescanning of `--format` option for the `dumpdata` command, and based on the prescanned format, calls `add_arguments()` on the format to register format-specific commandline arguments.
 - Lets the 'yaml' serializer register a `--yaml-collection-style` option. This behaves the same as before, except that it is now yaml-specific.
 - Modifies the dumpdata test code to allow passing the new option, without *always* passing it (since other formats will note accept it).

The problem I'm still facing is with the test suite. I had added [testcases for dumpdata](https://github.com/django/django/pull/10117/files#diff-14b4883f2cf50c9daf38f2b4ff9224acR653) that passed the new `yaml_collection_style` option. These testcases use [`django.core.management.call_command`](https://github.com/django/django/blob/0d031f9d40c8dfeb57a233501707a6f4bbff3987/django/core/management/__init__.py#L75) helper function, which accepts *parsed commandline options*, rather than the actual commandline. This means it does not call [`Command.run_from_argv`](https://github.com/django/django/blob/0d031f9d40c8dfeb57a233501707a6f4bbff3987/django/core/management/commands/dumpdata.py#L25-L31) to parse the commandline, bypassing the prescanning. However, it *does* create a commandline parser and uses it to figure out what options would be valid. Since the YAML-specific options are not present in this parser (since no prescanning happened), the `yaml_collection_style` option will always be rejected.

I'm not sure what the best way forward is, but I can imagine three ways to fix this:
 - Modify the testcase to actually run the command, rather than bypassing the fork and commandline parsing using `call_command`. This is what the tests for `manage.py test --testrunner` do, which is why this is not a problem there.
 - Modify the management commands (in particular [`call_command`](https://github.com/django/django/blob/0d031f9d40c8dfeb57a233501707a6f4bbff3987/django/core/management/__init__.py#L75)  and [`ManagementUtility.execute()`](https://github.com/django/django/blob/0d031f9d40c8dfeb57a233501707a6f4bbff3987/django/core/management/__init__.py#L301)) to explicitly support this prescanning (rather than making it a detail inside the `test` and `dumpdata` commands). This would mean extending the command-API with something like `add_prescan_arguments()` (or perhaps call `add_arguments()` twice), using argparse's `parse_known_args` to parse just those, and only then call the real `add_arguments()`. I'm not sure how to pass the prescanned arguments to `add_arguments()`, though, since the current API does not specify arbitrary kwargs. This will probably need an extra method, `process_prescanned_arguments()` or something. I can make this work, but I can't see this being really elegant.
 - Drop the prescanning, and instead  use a custom argparse `Action` class. This could allow a single argument-parsing pass, where, once the `--format` option is parsed, calls `add_arguments` on the specified format, modifying the parser during the parsing process. I'm not entirely sure if this works, and would be a supported usecases for argparse, but it does sound elegant. This would mean that any format-specific options would need to appear *after* `--format`, but that seems reasonable enough. It would also mean that the options for the default format would need to be removed again, which [might not be possible without digging further into argparse internals](https://bugs.python.org/issue19462) than we're already doing (and what would happen when you put json-specific options on the commandline followed by `--format=yaml`?). Also, this would need changes in `call_command`, which now digs into argparse internals to figure out available options, which might not work with this approach...

Conceptually, the last option seems the most elegant, except that I'm afraid that argparse does not really support this case (especially wrt options related to the default format). I just came up with this option while writing this comment, so I haven't really tried anything yet.

Any other insights?

matthijskooijman commented:
@timgraham (and others), any insights in the best way to fix this (I outlined a few possibilities in the previous comment). I haven't looked at this since then, but if some consensus is reached on the way forward, I can see if I can move this forward again.

carltongibson commented:
Closing as per ticket.
